### ISSUE TITLE
singledispatch handlers falsely reported as not accessed

### ISSUE THREAD

**Issue by jackric** [microsoft/pyright#1166]
**Describe the bug**
Using `functools.singledispatch` with `reportUnusedFunction` turned on causes false reports about unused functions for the registered handlers.

**To Reproduce**
pyrightconfig.json
```json
{
  "reportUnusedFunction": true
}
```
example.py
```python
from functools import singledispatch

@singledispatch
def handle(thing):
    pass

@handle.register(int)
def _handle_int(thing: int):
    print(f"The number {thing}")

@handle.register(str)
def _handle_str(thing: str):
    print(f"The string {thing}")

handle(55)
handle("spam")
```
Current output
```
example.py
  8:5 - error: Function "_handle_int" is not accessed (reportUnusedFunction)
  12:5 - error: Function "_handle_str" is not accessed (reportUnusedFunction)
```

**Expected behavior**
Don't report the decorated functions as not accessed. I think there are 2 popular categories of decorators to consider: what I'll call "pure" decorators which just augment and return the decorant, and then "registration" decorators like this case. Perhaps registration decorators can be annotated so that their decorants aren't reported as unused.

I am aware of the workaround here of removing the underscore prefix on the functions but these handlers aren't meant to be accessed outside their module!

**VS Code extension or command-line**
Command line

**Comment by erictraut** [comment]
These symbols are not accessed in any way that is visible to the static type checker, and they are marked as private (names start with an underscore). There's no provision in the Python type language to annotate decorators as being "registrants".

The workaround is to remove the underscore if the grayed-out text really bothers you. You can suppress the `reportUnusedFunction` diagnostic for this file only using `# pyright: reportUnusedFunction=false`.

**Comment by jackric** [comment]
Thanks for the guidance, I'll probably adopt the per-file config override for my uses.

> There's no provision in the Python type language to annotate decorators as being "registrants"

Fair point, I think a proposal for this in Python has legs. Many web apps register handlers via decorator.
Does the Pyright team feed ideas into the Python typing conversation?
